src/filesystem.cc

```diff
@@ -1473,22 +1507,46 @@ S3FileSystem::S3FileSystem(
     }
   }
 
-  if ((secret_key != NULL) && (key_id != NULL)) {
-    client_ = s3::S3Client(
+  if (secret_key.size() && key_id.size()) {
+    client_ = new s3::S3Client(
         credentials, config,
         Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::Never,
         /*useVirtualAdressing*/ false);
 
   } else {
-    client_ = s3::S3Client(
+    client_ = new s3::S3Client(
         config, Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::Never,
         /*useVirtualAdressing*/ false);
   }
 }
 
 S3FileSystem::~S3FileSystem()
 {
-  Aws::ShutdownAPI(options_);
+  // should delete all client_ before shutdown
+  delete client_;
+  // shutdown aws api if no more S3 instances
+  // FIXME: Calling Aws::ShutdownAPI will result in segmentation fault.
+  //        Currently, the Aws::initAPI will be called at most once at the
+  //        constructor and zero calls to Aws::ShutdownAPI, which is not a
+  //        memory leak and resources will be reclaimed when triton stops.
+  /*
+  std::lock_guard<std::mutex> lock(init_mu_);
+  if (!--init_count_) {
+    Aws::ShutdownAPI(options_);
+  }
+  */
```

Comment:
remove if no longer needed

Reply:
Yes. Moved to smart pointer for `client_`, and no more destructors.